### ISSUE TITLE
[Update] Cloud Firestore Xcode Build Times

### ISSUE THREAD

**Issue by Salakar** [FirebaseExtended/flutterfire#2751]
Hey everyone 👋 

One of the key issues we wanted to look into providing a solution for was https://github.com/FirebaseExtended/flutterfire/issues/349 - improving build times for Xcode when using Firestore.

I'm happy to say we now have something for you to try (please provide feedback) that should significantly improve your build times;

**Mac mini (2018) 6 cores**:

```
Before:    ~ 240s
After:     ~  45s
```

**GitHub Action CI 2 cores**:

```
Before:    ~ 551s
After:     ~ 174s
```

----

### Why is it slow?

Currently the Firestore iOS SDK depends on some 500k lines of mostly C++ code (from dependencies such as gRPC), which when compiling as part of your Xcode build can take a long time - even more so in CI environments.

### How can we speed it up?

The biggest win to improving build times would be to not have to compile all that C++ code. The way to do that would be to consume precompiled binaries, in the new `XCFramework` bundle type introduced to Xcode last year, luckily as part of the Firebase iOS SDK release process these bundles are generated and published in a zip containing all Firebase SDKs.

To help you consume these quicker with the least amount of changes to your codebase we've created [invertase/firestore-ios-sdk-frameworks](https://github.com/invertase/firestore-ios-sdk-frameworks) - a repository that provides a drop in replacement podspec for Firestore that consumes the precompiled binaries instead, without you needing to download 3gb of SDKs and manually extract out the Firestore SDKs from the rest of the SDKs in the zip.

#### ⚡ Setup ⚡

To integrate, you just need to add 1 line to your `ios/Podfile` inside your Flutter project;

```ruby
pod 'FirebaseFirestore', :git => 'https://github.com/invertase/firestore-ios-sdk-frameworks.git', :tag => '7.4.0'
```

Add this line inside your `target 'Runner' do` block in your Podfile, e.g.:

```ruby
# ...
target 'Runner' do
  pod 'FirebaseFirestore', :git => 'https://github.com/invertase/firestore-ios-sdk-frameworks.git', :tag => '7.4.0'

# ...
end
```

No changes to the FlutterFire packages were required. See #4698 for a full example of integrating with automatic versioning based on what `firebase_core` uses.

### Why not ship the binaries with the pub packages?

This would add ~200mb to every release. Whereas via a custom pod this is only ever downloaded once on your machine per Firebase iOS SDK version.

----

For more information on the `pod` and versioning please see the [invertase/firestore-ios-sdk-frameworks](https://github.com/invertase/firestore-ios-sdk-frameworks) repository.

----

**⚠️ Note:** if you were previously caching iOS builds on CI you may now find that when using precompiled binaries that caching is no longer required and it may actually slow down your build times by several minutes. See this comment below for more; https://github.com/FirebaseExtended/flutterfire/issues/2751#issuecomment-643190500

----

✨ **Please let me know if this has helped you, and I would love to see some before and after timings if you feel like sharing** ✨

**Comment by mikehardy** [comment]
This actually works for more than flutter, for react-native-firebase or anyone that integrates the firestore pod it works, I know this because my creaky old 4 core mbp takes about 5mins less to compile and get the app running. 1 line! Thank you.

**Comment by Salakar** [comment]
> This actually works for more than flutter, for react-native-firebase or anyone that integrates the firestore pod it works, I know this because my creaky old 4 core mbp takes about 5mins less to compile and get the app running. 1 line! Thank you.

Yep, works for anyone using Firestore on iOS - no matter the framework 🎉

**Comment by jsoms** [comment]
Thanks for the tip.
However when I try to implement it, I get :
 [!] CocoaPods could not find compatible versions for pod "FirebaseFirestore":
      In snapshot (Podfile.lock):
        FirebaseFirestore (= 1.11.2, ~> 1.11.2)
      In Podfile:
        FirebaseFirestore (from `https://github.com/invertase/firestore-ios-sdk-frameworks.git`, tag `6.26.0`)
    None of your spec sources contain a spec satisfying the dependencies: `FirebaseFirestore (from `https://github.com/invertase/firestore-ios-sdk-frameworks.git`, tag `6.26.0`), FirebaseFirestore (= 1.11.2, ~> 1.11.2)`.
    You have either:
     * out-of-date source repos which you can update with `pod repo update` or with `pod install --repo-update`.
     * mistyped the name or version.
     * not added the source repo that hosts the Podspec to your Podfile.

I tried to pod update, but with no luck

Any idea ?

**Comment by Ehesp** [comment]
@jsoms try removing your `Podfile.lock` and trying again.

**Comment by orestesgaolin** [comment]
I started simple tests on codemagic and there's at least few minutes decrease in iOS build time. Unfortunately cache upload caused the total build to be still rather long.

I'm going to update this table with next results.


Flutter stable 1.17.3
Xcode 11.5


|  Id | Precompiled binaries | Cache | cloud_firestore version | Firestore |   Fetch | Installing dep |     iOS | pod install | Xcode build |   Total |
|----:|---------------------:|------:|-------------------------|-----------|--------:|---------------:|--------:|------------:|------------:|--------:|
| 251 |                   no |   yes | 0.13.4+2                | 6.23.0    | 0:03:10 |        0:01:01 | 0:27:43 |         291 |         604 | 0:48:20 |
| 253 |                  yes |   yes | 0.13.4+2                | 6.23.0    | 0:04:28 |        0:01:27 | 0:16:17 |         193 |         402 | 0:42:00 |
|     |                      |       |                         |           |         |                |         |             |             |         |

**Comment by Salakar** [comment]
@orestesgaolin what's it like without uploading build caches?

**Comment by orestesgaolin** [comment]
@Salakar I updated my comment above and it's so surprising to see that without any cache (i.e. completely clean build on new workflow) the build time dropped dramatically to about 13 min. I'm gonna run it again without any cache

**Comment by orestesgaolin** [comment]
@Salakar second build **without cache** was a bit longer, so it seems that it doesn't have significant effect (thus it may be completely unnecessary to use). I will be collecting data in subsequent builds and maybe we can then draw some conclusions out of it

**Comment by jmagman** [comment]
Note using `xcframeworks` requires CocoaPods 1.9.0 or higher.  Flutter currently requires a minimum of 1.6.

@Salakar Created issue to add `cocoapods_version` to the podspec. https://github.com/invertase/firestore-ios-sdk-frameworks/issues/1.

**Comment by ElaineSchwner** [comment]
Could you help me?
- I run 'pod deintegrate';
- Deleted Pofile.lock
- Run 'pod install' 

And I got this:
[!] CocoaPods could not find compatible versions for pod "FirebaseFirestore":
  In Podfile:
    FirebaseFirestore (from `https://github.com/invertase/firestore-ios-sdk-frameworks.git`, tag `6.26.0`)

    cloud_firestore (from `.symlinks/plugins/cloud_firestore/ios`) was resolved to 0.0.1, which depends on
      Firebase/Firestore (~> 6.0) was resolved to 6.0.0, which depends on
        FirebaseFirestore (~> 1.3.0)

What must I do?
Thank you.

**Comment by red42** [comment]
@ElaineSchwner I solved a similar problem with `pod repo update` as per @jsoms ' comment.

**Comment by ElaineSchwner** [comment]
> pod repo update

Solved !
Thank you!

**Comment by rafekun** [comment]
It worked well. Thanks a lot 🛩️ 😄

**Comment by RaviKavaiya** [comment]
This didn't work for me. I tried `pod install`, `pod update`...
See this:
```
Xcode build done.
Could not build the application for the simulator.
Error launching application on iPhone 11.
```

**Comment by Ehesp** [comment]
Hi @RaviKavaiya we'll need more information to help on that one. It looks like the XCode build succeeded but then fails when the app is launched - what are the errors?

**Comment by RaviKavaiya** [comment]
@Ehesp I tried running with flutter run -v. Couldn't find anything. Will give another shot later and will tell you..

Thanks

**Comment by andre-arsenault** [comment]
Build timings on my MacBook Pro (2018) 4 cores:

```
Before:   ~ 272s
After:    ~ 122s
```
A very nice speed improvement, thank you!

How can we tell which version of the iOS SDK the `cloud_firestore` package uses, so we can keep the tag up to date in our Podfile? I hope that will be part of the package's release notes going forward?

**Comment by rhisiart-cufflink** [comment]
>ld: library not found for -lBoringSSL-GRPC

I've raised an [issue](https://github.com/invertase/firestore-ios-sdk-frameworks/issues/7) against the `firestore-ios-sdk-frameworks` library - referencing it here in case it helps someone else...

**Comment by farukozsoy** [comment]
On an old Macbook Pro ( 2011 - 2 cores - i7 )

**before ~ 15 minutes
after ~ 9 minutes**

Thanks!

**Comment by AndroidNils** [comment]
Is there also a workaround for Flutter macOS?

**Comment by Salakar** [comment]
@AndroidNils unfortunately this is only for iOS

**Comment by luckyhandler** [comment]
I have an issue when trying to build this for production:

```
Xcode's output:
↳
    .../ios/Runner/GeneratedPluginRegistrant.m:10:9: fatal error: module 'cloud_firestore' not found
    @import cloud_firestore;
     ~~~~~~~^~~~~~~~~~~~~~~
    1 error generated.
    .../ios/Runner/GeneratedPluginRegistrant.m:10:9: fatal error: module 'cloud_firestore' not found
    @import cloud_firestore;
     ~~~~~~~^~~~~~~~~~~~~~~
    1 error generated.
    note: Using new build system
    note: Building targets in parallel
    note: Planning build
    note: Constructing build description
```
anyone else having this issue when running `flutter build ios`?

**Comment by deondejongh** [comment]
Any ideas as to how to implement this suggested solution when one doesn't have access to a Mac? I'm using VSCode and Codemagic for builds, so not sure how to edit Podfile?

**Comment by FindtheTreasure** [comment]
After upgrading my pod Firestore to 6.26.0 I now get a build fail on iOS:
`    /Users/garethbeall/Developer/flutter/.pub-cache/hosted/pub.dartlang.org/cloud_firestore-0.13.7/ios/Classes/FLTCloudFirestorePlugin.m:107:18: error: use of undeclared identifier 'fieldPath'; did you mean 'fieldName'?
          } else if (fieldPath != nil) {
                     ^~~~~~~~~
                     fieldName
    /Users/garethbeall/Developer/flutter/.pub-cache/hosted/pub.dartlang.org/cloud_firestore-0.13.7/ios/Classes/FLTCloudFirestorePlugin.m:91:15: note: 'fieldName' declared here
        NSString *fieldName = nil;
                  ^
    fatal error: too many errors emitted, stopping now [-ferror-limit=]
    20 errors generated.
    note: Using new build system
    note: Building targets in parallel
    note: Planning build
    note: Constructing build description
Could not build the precompiled application for the device.

Error launching application on Physical Device.`

Any help would be greatly appreciated.

I have tried:
- flutter clean
- deleting the podfile.lock
- pod repo update
- pod update

**Comment by sijangurung** [comment]
Nice !!!
Macbook pro 2,9 GHz Quad-Core Intel Core i7. 16 GB ram
before: Xcode build done.                                        223.0s
after: Xcode build done.                                           83.1s

for flutter -run  👍 💯 👍

**Comment by Jman1080** [comment]
Tried with a new `futter create app` and the build fails with 20 errors

**pubspec.yaml:** 

```
environment:
  sdk: ">=2.7.0 <3.0.0"

dependencies:
  flutter:
    sdk: flutter
  cloud_firestore: ^0.14.0
```

**output:** 

```
Xcode build done.                                           32.2s
Failed to build iOS app
Error output from Xcode build:
↳
    ** BUILD FAILED **


Xcode's output:
↳
    In file included from /Users/jay/Documents/SDKs/flutter/.pub-cache/hosted/pub.dartlang.org/cloud_firestore-0.14.0/ios/Classes/FLTFirebaseFirestoreReader.m:9:
    /Users/jay/Documents/SDKs/flutter/.pub-cache/hosted/pub.dartlang.org/cloud_firestore-0.14.0/ios/Classes/Private/FLTFirebaseFirestoreUtils.h:42:4: error: expected a type
    + (FIRFirestoreSource)FIRFirestoreSourceFromArguments:(NSDictionary *_Nonnull)arguments;
       ^
    /Users/jay/Documents/SDKs/flutter/.pub-cache/hosted/pub.dartlang.org/cloud_firestore-0.14.0/ios/Classes/Private/FLTFirebaseFirestoreUtils.h:44:4: error: expected a type
    + (FIRFirestore *_Nullable)getCachedFIRFirestoreInstanceForKey:(NSString *_Nonnull)key;
       ^
    /Users/jay/Documents/SDKs/flutter/.pub-cache/hosted/pub.dartlang.org/cloud_firestore-0.14.0/ios/Classes/Private/FLTFirebaseFirestoreUtils.h:45:40: error: expected a type
    + (void)setCachedFIRFirestoreInstance:(FIRFirestore *_Nonnull)firestore
                                           ^
    /Users/jay/Documents/SDKs/flutter/.pub-cache/hosted/pub.dartlang.org/cloud_firestore-0.14.0/ios/Classes/FLTFirebaseFirestoreReader.m:92:4: error: expected a type
    - (FIRFirestoreSettings *)FIRFirestoreSettings {
       ^
    /Users/jay/Documents/SDKs/flutter/.pub-cache/hosted/pub.dartlang.org/cloud_firestore-0.14.0/ios/Classes/FLTFirebaseFirestoreReader.m:122:4: error: expected a type
    - (FIRQuery *)FIRQuery {
       ^
    /Users/jay/Documents/SDKs/flutter/.pub-cache/hosted/pub.dartlang.org/cloud_firestore-0.14.0/ios/Classes/FLTFirebaseFirestoreReader.m:215:4: error: expected a type
    - (FIRFirestore *)FIRFirestore {
       ^
    /Users/jay/Documents/SDKs/flutter/.pub-cache/hosted/pub.dartlang.org/cloud_firestore-0.14.0/ios/Classes/FLTFirebaseFirestoreReader.m:25:16: error: use of undeclared identifier 'FIRTimestamp'
          return [[FIRTimestamp alloc] initWithSeconds:seconds nanoseconds:nanoseconds];
                   ^
    /Users/jay/Documents/SDKs/flutter/.pub-cache/hosted/pub.dartlang.org/cloud_firestore-0.14.0/ios/Classes/FLTFirebaseFirestoreReader.m:33:16: error: use of undeclared identifier 'FIRGeoPoint'
          return [[FIRGeoPoint alloc] initWithLatitude:latitude longitude:longitude];
                   ^
    /Users/jay/Documents/SDKs/flutter/.pub-cache/hosted/pub.dartlang.org/cloud_firestore-0.14.0/ios/Classes/FLTFirebaseFirestoreReader.m:36:7: error: use of undeclared identifier 'FIRFirestore'
          FIRFirestore *firestore = [self readValue];
          ^
    /Users/jay/Documents/SDKs/flutter/.pub-cache/hosted/pub.dartlang.org/cloud_firestore-0.14.0/ios/Classes/FLTFirebaseFirestoreReader.m:36:21: error: use of undeclared identifier 'firestore'
          FIRFirestore *firestore = [self readValue];
                        ^
    /Users/jay/Documents/SDKs/flutter/.pub-cache/hosted/pub.dartlang.org/cloud_firestore-0.14.0/ios/Classes/FLTFirebaseFirestoreReader.m:38:15: error: use of undeclared identifier 'firestore'
          return [firestore documentWithPath:documentPath];
                  ^
    /Users/jay/Documents/SDKs/flutter/.pub-cache/hosted/pub.dartlang.org/cloud_firestore-0.14.0/ios/Classes/FLTFirebaseFirestoreReader.m:47:16: error: use of undeclared identifier 'FIRFieldPath'
          return [[FIRFieldPath alloc] initWithFields:array];
                   ^
    /Users/jay/Documents/SDKs/flutter/.pub-cache/hosted/pub.dartlang.org/cloud_firestore-0.14.0/ios/Classes/FLTFirebaseFirestoreReader.m:52:15: error: use of undeclared identifier 'FIRFieldValue'
          return [FIRFieldValue fieldValueForArrayUnion:[self readValue]];
                  ^
    /Users/jay/Documents/SDKs/flutter/.pub-cache/hosted/pub.dartlang.org/cloud_firestore-0.14.0/ios/Classes/FLTFirebaseFirestoreReader.m:54:15: error: use of undeclared identifier 'FIRFieldValue'
          return [FIRFieldValue fieldValueForArrayRemove:[self readValue]];
                  ^
    /Users/jay/Documents/SDKs/flutter/.pub-cache/hosted/pub.dartlang.org/cloud_firestore-0.14.0/ios/Classes/FLTFirebaseFirestoreReader.m:56:15: error: use of undeclared identifier 'FIRFieldValue'
          return [FIRFieldValue fieldValueForDelete];
                  ^
    /Users/jay/Documents/SDKs/flutter/.pub-cache/hosted/pub.dartlang.org/cloud_firestore-0.14.0/ios/Classes/FLTFirebaseFirestoreReader.m:58:15: error: use of undeclared identifier 'FIRFieldValue'
          return [FIRFieldValue fieldValueForServerTimestamp];
                  ^
    /Users/jay/Documents/SDKs/flutter/.pub-cache/hosted/pub.dartlang.org/cloud_firestore-0.14.0/ios/Classes/FLTFirebaseFirestoreReader.m:61:12: error: use of undeclared identifier 'FIRFieldValue'
              [FIRFieldValue fieldValueForDoubleIncrement:((NSNumber *)[self readValue]).doubleValue];
               ^
    /Users/jay/Documents/SDKs/flutter/.pub-cache/hosted/pub.dartlang.org/cloud_firestore-0.14.0/ios/Classes/FLTFirebaseFirestoreReader.m:63:15: error: use of undeclared identifier 'FIRFieldValue'
          return [FIRFieldValue fieldValueForIntegerIncrement:((NSNumber *)[self readValue]).intValue];
                  ^
    /Users/jay/Documents/SDKs/flutter/.pub-cache/hosted/pub.dartlang.org/cloud_firestore-0.14.0/ios/Classes/FLTFirebaseFirestoreReader.m:65:15: error: use of undeclared identifier 'FIRFieldPath'
          return [FIRFieldPath documentID];
                  ^
    fatal error: too many errors emitted, stopping now [-ferror-limit=]
    20 errors generated.
    note: Using new build system
    note: Building targets in parallel
    note: Planning build
    note: Constructing build description

```

**Comment by rubenstolk** [comment]
You'll need to do some cleaning: https://github.com/FirebaseExtended/flutterfire/issues/2507#issuecomment-625282792

**Comment by esDotDev** [comment]
Testing with a hello world, using a 2 core Mac Mini as the builder.
Before: 325s
After 39s 

This should really be called out on the front page. I didn't realize this was caused by firebase, and wasted a lot of time before finding this post tucked away. 

With that said, thanks for the fix!!

**Comment by Salakar** [comment]
> Testing with a hello world w/ a bunch of plugins, using a 2 core Mac Mini as the builder.
> Before: 325s
> After 39s
> 
> This should really be called out on the front page. I didn't realize this was caused by firebase, and wasted a lot of time before finding this post tucked away.
> 
> With that said, thanks for the fix!!

If you're willing @esDotDev, could you submit a PR to add it as an optional step for iOS here: https://firebase.flutter.dev/docs/firestore/overview

Theres an edit button on the bottom of every page on the docs site. 🤓

**Comment by esDotDev** [comment]
That's cool! I don't have time atm but I think we probably want to stick it here: 
https://firebase.flutter.dev/docs/installation/ios

**Comment by esDotDev** [comment]
Man at this point I just don't know what to do. So completely frustrated and at a wall with 2 days and no progress at all. 

I'm just trying to run the Counter App, with firebase_core and cloud_firestore, and it just refuses to work. 
```
version: 1.0.0+1

environment:
  sdk: ">=2.7.0 <3.0.0"

dependencies:
  flutter:
    sdk: flutter

  cloud_firestore: ^0.14.0+2
  firebase_core: ^0.5.0
  path_provider: ^1.6.14
```

I've tried every version of Flutter, Stable, Beta, Master, multiple specific versions
I've cleaned everything, so many times I've lost count, pub cache, Podfile, Pod lock, Pod Cache, you name it, I've deleted it 
I've built over 50 times, just will not compile with cloud_firestore
I've tried both Switch and ObjC versions


```
dependencies:
  flutter:
    sdk: flutter

  cloud_firestore: ^0.14.0+2
  firebase_core: ^0.5.0
```
Always with these same errors:
```
 /Users/home/Development/_sdks/flutter/.pub-cache/hosted/pub.dartlang.org/cloud_firestore-0.14.0+2/ios/Classes/FLTFirebaseFirestoreWriter.m:20:5: error: unknown type name 'FIRTimestamp'; did you mean 'CVTimeStamp'?
        FIRTimestamp *timestamp = value;
        ^~~~~~~~~~~~
        CVTimeStamp
    In module 'UIKit' imported from /Users/home/Development/spikes/flutter_plugins_objc/ios/Pods/Target Support Files/cloud_firestore/cloud_firestore-prefix.pch:2:
    In module 'CoreImage' imported from
    /Applications/Xcode.app/Contents/Developer/Platforms/iPhoneSimulator.platform/Developer/SDKs/iPhoneSimulator13.6.sdk/System/Library/Frameworks/UIKit.framework/Headers/UIColor.h:13:
    In module 'CoreVideo' imported from
    /Applications/Xcode.app/Contents/Developer/Platforms/iPhoneSimulator.platform/Developer/SDKs/iPhoneSimulator13.6.sdk/System/Library/Frameworks/CoreImage.framework/Headers/CIImage.h:10:
    /Applications/Xcode.app/Contents/Developer/Platforms/iPhoneSimulator.platform/Developer/SDKs/iPhoneSimulator13.6.sdk/System/Library/Frameworks/CoreVideo.framework/Headers/CVBase.h:309:3: note: 'CVTimeStamp'
    declared here
    } CVTimeStamp; 
      ^
    /Users/home/Development/_sdks/flutter/.pub-cache/hosted/pub.dartlang.org/cloud_firestore-0.14.0+2/ios/Classes/FLTFirebaseFirestoreWriter.m:20:31: error: implicit conversion of Objective-C pointer type 'id' to C
    pointer type 'CVTimeStamp *' requires a bridged cast
        FIRTimestamp *timestamp = value;
                                  ^~~~~
    /Users/home/Development/_sdks/flutter/.pub-cache/hosted/pub.dartlang.org/cloud_firestore-0.14.0+2/ios/Classes/FLTFirebaseFirestoreWriter.m:20:31: note: use __bridge to convert directly (no change in ownership)
        FIRTimestamp *timestamp = value;
                                  ^
                                  (__bridge CVTimeStamp *)( )
    /Users/home/Development/_sdks/flutter/.pub-cache/hosted/pub.dartlang.org/cloud_firestore-0.14.0+2/ios/Classes/FLTFirebaseFirestoreWriter.m:20:31: note: use CFBridgingRetain call to make an ARC object available as a
    +1 'CVTimeStamp *'
        FIRTimestamp *timestamp = value;
                                  ^
                                  CFBridgingRetain( )
    /Users/home/Development/_sdks/flutter/.pub-cache/hosted/pub.dartlang.org/cloud_firestore-0.14.0+2/ios/Classes/FLTFirebaseFirestoreWriter.m:20:19: warning: incompatible pointer types initializing 'CVTimeStamp *'
    with an expression of type '__strong id' [-Wincompatible-pointer-types]
        FIRTimestamp *timestamp = value;
                      ^           ~~~~~
    /Users/home/Development/_sdks/flutter/.pub-cache/hosted/pub.dartlang.org/cloud_firestore-0.14.0+2/ios/Classes/FLTFirebaseFirestoreWriter.m:21:31: error: member reference type 'CVTimeStamp *' is a pointer; did you
    mean to use '->'?
        SInt64 seconds = timestamp.seconds;
                         ~~~~~~~~~^
                                  ->
    /Users/home/Development/_sdks/flutter/.pub-cache/hosted/pub.dartlang.org/cloud_firestore-0.14.0+2/ios/Classes/FLTFirebaseFirestoreWriter.m:21:32: error: no member named 'seconds' in 'CVTimeStamp'
        SInt64 seconds = timestamp.seconds;
                         ~~~~~~~~~ ^
    /Users/home/Development/_sdks/flutter/.pub-cache/hosted/pub.dartlang.org/cloud_firestore-0.14.0+2/ios/Classes/FLTFirebaseFirestoreWriter.m:22:32: error: member reference type 'CVTimeStamp *' is a pointer; did you
    mean to use '->'?
        int nanoseconds = timestamp.nanoseconds;
                          ~~~~~~~~~^
                                   ->
    /Users/home/Development/_sdks/flutter/.pub-cache/hosted/pub.dartlang.org/cloud_firestore-0.14.0+2/ios/Classes/FLTFirebaseFirestoreWriter.m:22:33: error: no member named 'nanoseconds' in 'CVTimeStamp'
        int nanoseconds = timestamp.nanoseconds;
                          ~~~~~~~~~ ^
    /Users/home/Development/_sdks/flutter/.pub-cache/hosted/pub.dartlang.org/cloud_firestore-0.14.0+2/ios/Classes/FLTFirebaseFirestoreWriter.m:26:36: error: use of undeclared identifier 'FIRGeoPoint'
      } else if ([value isKindOfClass:[FIRGeoPoint class]]) {
                                       ^
    /Users/home/Development/_sdks/flutter/.pub-cache/hosted/pub.dartlang.org/cloud_firestore-0.14.0+2/ios/Classes/FLTFirebaseFirestoreWriter.m:27:5: error: use of undeclared identifier 'FIRGeoPoint'
        FIRGeoPoint *geoPoint = value;
        ^
    /Users/home/Development/_sdks/flutter/.pub-cache/hosted/pub.dartlang.org/cloud_firestore-0.14.0+2/ios/Classes/FLTFirebaseFirestoreWriter.m:27:18: error: use of undeclared identifier 'geoPoint'
        FIRGeoPoint *geoPoint = value;
                     ^
    /Users/home/Development/_sdks/flutter/.pub-cache/hosted/pub.dartlang.org/cloud_firestore-0.14.0+2/ios/Classes/FLTFirebaseFirestoreWriter.m:28:24: error: use of undeclared identifier 'geoPoint'
        Float64 latitude = geoPoint.latitude;
                           ^
    /Users/home/Development/_sdks/flutter/.pub-cache/hosted/pub.dartlang.org/cloud_firestore-0.14.0+2/ios/Classes/FLTFirebaseFirestoreWriter.m:29:25: error: use of undeclared identifier 'geoPoint'
        Float64 longitude = geoPoint.longitude;
                            ^
    fatal error: too many errors emitted, stopping now [-ferror-limit=]
    1 warning and 20 errors generated.
    note: Using new build system
    note: Building targets in parallel
    note: Planning build
    note: Constructing build description

Could not build the application for the simulator.
Error launching application on iPhone SE (2nd generation).
```

**Comment by Fonz001** [comment]
For some reason, the proposed update to the `Podfile` would only work with `CocoaPods 1.9.1` (and higher probably)

Not sure if it's related to the `Podfile` update or the used plugins itself, but leaving this here for people who are having a hard time to get this to work.

**Comment by esDotDev** [comment]
> For some reason, the proposed update to the `Podfile` would only work with `CocoaPods 1.9.1` (and higher probably)
> 
> Not sure if it's related to the `Podfile` update or the used plugins itself, but leaving this here for people who are having a hard time to get this to work.

It seems like this might have been the issue. You just saved me a ton of head to wall smashing, thanks a mill!

**Comment by esDotDev** [comment]
> If you're willing @esDotDev, could you submit a PR to add it as an optional step for iOS here: https://firebase.flutter.dev/docs/firestore/overview

Thought it made more sense here so it can't get missed, and it's around the area we edit the podfile for other reasons. 
https://github.com/FirebaseExtended/flutterfire/pull/3365

**Comment by AndroidNils** [comment]
@Salakar Is there any workaround for Flutter macOS? I opened a new issue for Flutter macOS and Cloud Firestore: #3185

**Comment by esDotDev** [comment]
Sorry, looks like I spoke to soon. When using the pre-compiled binary I get 2 errors.

When using `firebase_core` alone, it complains 
` ld: framework not found BoringSSL-GRPC.xcframework`

When using `firebase_core` plus `cloud_firestore` it blows up with random missing declaration errors, like: 
`/Users/home/Development/_sdks/flutter/.pub-cache/hosted/pub.dartlang.org/cloud_firestore-0.14.0+2/ios/C
    lasses/FLTFirebaseFirestorePlugin.m:42:54: error: unknown type name 'FIRTransaction'; did you mean
    'CATransaction'?`

When removing the binary, and just trying to use `firebase_core` plus `cloud_firestore` I get like 300 lines about explaining some compatability matrix I couldn't possible begin to parse.

**Comment by esDotDev** [comment]
Interestingly toasting the entire build folder and rebuilding, it now compiles and launches ok. 

But I had just previously done that, and the only change I had made was to include the binary, and then remove it, which seems to have left the iOS install in a totally broken state. I did try all the obvious stuff like flutter clean, and deleting the pod lock, and cleaning the Xcode build, but didn't seem to help.

**Comment by mikehardy** [comment]
I know this is flutter and not react-native but I'm listening in on this issue because I help Invertase from time to time and the firestore pre-compiled framework is shared between FlutterFire and react-native-firebase

That said, if there is not some command line tooling to wipe out the various build caches and intermediate artifacts, someone in Flutter-land would earn lots of virtual beers by coding one up.

For reference / inspiration, this is the `react-native-clean-project` set of commands - *all* of which are needed at various times in order to truly reproduce clean builds. I run them all in CI for instance, and I never have bizarre problems, and the number of times I recommend it to users in react-native-firebase issues during triage and they then have success is quite high

https://github.com/pmadruga/react-native-clean-project/blob/master/source/internals/tasks.js

Cheers

**Comment by lukas2** [comment]
This is great, thanks guys! I am using this with Cordova. Mentioning it explicitly so Cordova programmers will find it quicker. 😇

**Comment by Salakar** [comment]
> This is great, thanks guys! I am using this with Cordova. Mentioning it explicitly so Cordova programmers will find it quicker. 😇

Great to hear this is benefitting other platforms too, so far Flutter, React Native & Cordova users reported using this with success :)

**Comment by realappie** [comment]
Thank you so much! For me this meant a reduction from about ~ 60 minutes to 10 minutes on bitrise.

**Comment by esDotDev** [comment]
Has no one on the team been able to reproduce the issue with missing Type Declarations? I'm super hesitant to re-enable the binary as I have everything working properly, and previously it borked the project.

I think we do need a good clean script, Parsing the example from @mikehardy and #issuecomment-625282792 we need something like: 

* Delete $FLUTTER_ROOT/.pub-cache
* Delete $HOME/Library/Caches/CocoaPods
* Delete ios/Pods
* Delete ios/build
* Run (killall Xcode || true) && xcrun -k   <-- not sure what this does
* Run xcodebuild -alltargets clean
* Run rm -rf "$(getconf DARWIN_USER_CACHE_DIR)/org.llvm.clang/ModuleCache"
* Run rm -rf "$(getconf DARWIN_USER_CACHE_DIR)/org.llvm.clang.$(whoami)/ModuleCache"
* Run rm -fr ~/Library/Developer/Xcode/DerivedData/ 
* Run rm -fr ~/Library/Caches/com.apple.dt.Xcode/'

Am missing anything?

**Comment by esDotDev** [comment]
Just tried adding the binary, same missing class exception errors, clean script doesn't seem to solve it :'(

**Comment by mikehardy** [comment]
The Xcode kill and simulator stop are needed (or...beneficial, the whole thing is a hack right? though it makes things 100% reproducible which is nice) because Xcode gets very upset when you delete the things those commands delete while it is still running, same with Simulator. So we get in front of that and just take them down.

**Comment by nerder** [comment]
Question: adding this will generate new files under`ios/build`. This should be versioned or `.gitignore`d ?

**Comment by mikehardy** [comment]
It is my understanding that *nothing* under ios/build should ever be versioned?

"build" implies (and is correct in this case) that they are derived/built from source, they are not themselves source. Nothing derived should be versioned?

**Comment by karimone** [comment]
Could someone help me with how to determine the correct tag version based on the packages I have? for example, I'm using:

```
  firebase_core: ^0.4.5
  firebase_auth: ^0.16.1
  firebase_storage: ^3.1.6
  firebase_messaging: ^6.0.16
  cloud_firestore: ^0.13.7
```

What version I should put in the runner?

**Comment by sultanmyrza** [comment]
> Could someone help me with how to determine the correct tag version based on the packages I have? for example, I'm using:
> 
> ```
>   firebase_core: ^0.4.5
>   firebase_auth: ^0.16.1
>   firebase_storage: ^3.1.6
>   firebase_messaging: ^6.0.16
>   cloud_firestore: ^0.13.7
> ```
> 
> What version I should put in the runner?

## Tested on fresh flutter project
When adding `cloud_firestore: ^0.14.0+2` build time is normal.
After adding `cloud_firestore: ^0.14.0+2` build time 142 seconds

After adding code from **snippet** in podfile 
1st build time 93 seconds
2nd build time 61.5 seconds


@karimone In your case maybe you can try to update cloud_firestore to lates version and add **snippet** from  in podfile
read more at https://firebase.flutter.dev/docs/overview/#improve-ios-build-times 

# Snippet
```
# ...
target `Runner' do
  pod 'FirebaseFirestore', :git => 'https://github.com/invertase/firestore-ios-sdk-frameworks.git', :tag => '6.26.0'
# ...
end
```
## I think only cloud_firestore slows down build time other firebase packages should not!

**Comment by sultanmyrza** [comment]
@karimone if you don't know which version to set
## before
```
# ...
target `Runner' do
  pod 'FirebaseFirestore', :git => 'https://github.com/invertase/firestore-ios-sdk-frameworks.git', :tag => '6.26.0'
# ...
end
```
Note: where the tag says 6.26.0 this should be changed to the pod version of Firebase/Firestore that you or your dependencies are using - in the format X.X.X, for FlutterFire the version that is being used can be seen here, for React Native Firebase here. If no version is specified on your current Firebase/Firestore pod then you can emit , :tag => '6.26.0' from the line above and use the latest version on master. 
you can read more [here](https://github.com/invertase/firestore-ios-sdk-frameworks#usage) 
## after
```
# ...
target `Runner' do
  pod 'FirebaseFirestore', :git => 'https://github.com/invertase/firestore-ios-sdk-frameworks.git'
# ...
end
```
### then `cd to ios`
### run `pod install --verbose`
### `--verbose` flag is important because `pod install` take some time so to be sure you are not stuck and waiting for nothing it's always good to see what is doing `pod install`

**Comment by jorg1piano** [comment]
> @AndroidNils unfortunately this is only for iOS

Is there a reason why the same solution only with a build for mac os wouldn't work? Long build times when introducing Firestore have been an issue for a really long time. I really like Firestore but spending hours building a project is not exactly cost effective. Any progress on this issue for macos?

**Comment by AndroidNils** [comment]
> > Some context as to why this doesn't work for macOS: invertase/firestore-ios-sdk-frameworks#11 (comment

**Comment by Etiene** [comment]
This change forced me to go on a rabbit hole of unrelated errors, but finally...
before: 404s
now: 168s

That's a big change, thanks!

**Comment by esDotDev** [comment]
I'm honestly done with Firebase due to this. The entire promise of easy cross-platform deploy is a joke. The entire build process is incredibly flaky on iOS. If you're stickin to Android only, then sure, but beyond that I don't see the pt.

**Comment by mikehardy** [comment]
:-), "basic socket connection" supporting encryption of course "to a NOSQL DB" in the cloud from a mobile device with flaky networking happening all the time and you'd like the socket to have all sorts of fun things like resumption, triggers from cloud to device that persist across network flakes etc, local caching.... Firebase or AWS or whoever, these aren't technical challenges to take lightly I don't think. The more features you support, the more code you need, and here we are

**Comment by esDotDev** [comment]
Certainly I oversimplified, but we have been doing all these things for years without 40m build times. Deeply regrettng my choice of firebase for chat, todo and img storage as deployment on iOS has been an absolute disaster. I spent literally 7 straight days, building 100s of times painfully slowly, and its still not fully resolved. Just wish google would be more upfront about deployment pains on IOS and quit pretending its smooth sailing when its really anything but.

**Comment by mulderpf** [comment]
This solution speeds up builds times, but it added almost 60MB to my IPA, so this is not really a feasible solution for me for doing production builds.  I've spent months wondering why things were so slow and really just not having loads of fun getting Flutter to build properly for iOS (I absolutely dread having to do anything which touches iOS - however, I mostly blame Apple, not Google!!)

**Comment by mikehardy** [comment]
@mulderpf I suspect something is very wrong with your project setup or something, when integrating this solution into react-native-firebase at least it has zero impact on final IPA size. My entire IPA is only 15MB or so, including this, so it is not this that causes the increase. The only time I have seen an increase like that is when I included some of the MLKit items that bundled machine learning models into the IPA

**Comment by mulderpf** [comment]
@mikehardy I'm not using react-native, I'm using Flutter and I'm using everything as standard apart from adding in the single suggested line.  Without it, my last successful build was 39.7MB (and took 44 mins 52 secs) and with it, my last successful build was 103.63MB (and took 17 mins 36 secs).  Not sure what could really be wrong as it is only a single line I've added in.

**Comment by mikehardy** [comment]
I understand it's Flutter not react-native-firebase, that is absolutely astounding though. It shouldn't happen and makes me think something else is wrong, like failure to do bitcode or strip symbols or something

**Comment by mulderpf** [comment]
Thanks @mikehardy - I decided to check again and now my "slow" build also created a huge IPA...not this then.  Did I mention how much I hate troubleshooting iOS stuff...give me Android any day.

**Comment by mikehardy** [comment]
I normally try to stay away from negative comments because a lot of times it's just my inexperience generating the bad result, but can bag on Xcode all day long ;-). Gradle has it's own issues but Xcode is next level painful yes. Glad to hear at least this speed hack wasn't the cause. The only problem I'm aware of with this now is that you may experience problems building from Xcode vs command line with Xcode12 and this integrated - or with this and Xcode12 and detox even from the command line, no triage yet https://github.com/invertase/firestore-ios-sdk-frameworks/issues/15

**Comment by esDotDev** [comment]
I think I may have finally figured out the issue with using the pre-compiled binary. I had run all the clean commands many times, with no luck.

In the end, it was this that did the trick: 
https://github.com/FirebaseExtended/flutterfire/issues/3424#issuecomment-685765415

Specifically:
1. Delete podfile.lock
2. Run `pod install --repo-update`

fwiw, here's the clean script I was using, it's pretty aggressive, rebuilding all of pub cache. I did run this before running the above.
```
#!/bin/zsh
echo "Clean Started..."
(killall Xcode || true)
xcrun -k
flutter clean
flutter pub upgrade
flutter pub get
flutter pub cache repair
rm -rf $FLUTTER_ROOT/.pub-cache
rm -rf $HOME/Library/Caches/CocoaPods
rm -rf ios/Pods
rm -rf ios/build
rm -rf "$(getconf DARWIN_USER_CACHE_DIR)/org.llvm.clang/ModuleCache"
rm -rf "$(getconf DARWIN_USER_CACHE_DIR)/org.llvm.clang.$(whoami)/ModuleCache"
rm -rf ~/Library/Developer/Xcode/DerivedData/
rm -rf ~/Library/Caches/com.apple.dt.Xcode/
echo "Clean Complete!"
```

**Comment by mikehardy** [comment]
@esDotDev oh - that's a shame, I have gone through that pain. At this point cocoapods commands to manage the podspec repo / lock file / update status are almost second nature. You are probably there now. This is pure-iOS development stuff, not at either the flutter or react-native level really (though obviously the cross-platform environment depends on the underlying...). Glad you found your problem, good luck with your project

**Comment by nerder** [comment]
Using the proposed fix when running `pod install --repo-update` i got the following error: 

```
[!] CocoaPods could not find compatible versions for pod "FirebaseFirestore":
  In Podfile:
    FirebaseFirestore (from `https://github.com/invertase/firestore-ios-sdk-frameworks.git`, tag `6.26.0`)

    cloud_firestore (from `.symlinks/plugins/cloud_firestore/ios`) was resolved to 0.14.2, which depends on
      Firebase/Firestore (~> 6.33.0) was resolved to 6.33.0, which depends on
        FirebaseFirestore (~> 1.18.0)
```

any help here?

**Comment by mikehardy** [comment]
@nerder - a couple things - 6.34.0 is out, I'd go for that if you can override to it?, but I would not go to v7 yet as that is a big change just this week and FlutterFire is not ready yet - in general for pod things just remove Podfile.lock every time you do pod operations, do a `pod repo update` to make sure you have new definitions, then `pod install`. That should clear things up.

You certainly can't mix the firestore-ios-sdk-frameworks reference at '6.26.0' into your main firebase-ios-sdk dependencies if they are set at '6.33.0', that won't work

**Comment by bounty1342** [comment]
@nerder, this will do :
```
# ...
target 'Runner' do
  pod 'FirebaseFirestore', :git => 'https://github.com/invertase/firestore-ios-sdk-frameworks.git', :tag => '6.33.0'

# ...
end
```

**Comment by luckyhandler** [comment]
I cannot download the precompiled libraries any more. When running `pod update` I get:
```
firebase_storage: Using user specified Firebase SDK version '7.0.0'
[!] CDN: trunk URL couldn't be downloaded: https://cdn.jsdelivr.net/cocoa/Specs/8/b/d/FirebaseCore/7.0.0/FirebaseCore.podspec.json Response: Timeout was reached
```
Any idea why?

**Comment by mikehardy** [comment]
@luckyhandler I have that problem when I rely on the cocoapods CDN sometimes. You might like this workaround:

https://github.com/mikehardy/rnfbdemo/blob/fc7e943ca95fe1777300985f31db2a075d2da4a5/make-demo.sh#L19-L21

**Comment by luckyhandler** [comment]
@mikehardy What fixes this issue is `pod repo remove trunk` and then `pod update`

**Comment by FetFrumos** [comment]
I use only firebase_messaging in my app. But time to build is very long(about 12-14 minutes). I try firestore-ios-sdk-frameworks, but I have many warnings  when build app and result error:

      `"std::__1::thread::hardware_concurrency()", referenced from:
          firebase::firestore::local::LevelDbRemoteDocumentCache::LevelDbRemoteDocumentCache(firebase::firestore::local::LevelDbPersistence*, firebase::firestore::local::LocalSerializer*) in
          FirebaseFirestore(leveldb_remote_document_cache.o)
      "___gxx_personality_v0", referenced from:
          -[FIRCollectionReference initWithPath:firestore:] in FirebaseFirestore(FIRCollectionReference.o)
          -[FIRCollectionReference initWithQuery:] in FirebaseFirestore(FIRCollectionReference.o)
          -[FIRCollectionReference isEqual:] in FirebaseFirestore(FIRCollectionReference.o)
          -[FIRCollectionReference isEqualToReference:] in FirebaseFirestore(FIRCollectionReference.o)
          absl::lts_2020_02_25::string_view::string_view<std::__1::allocator<char> >(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&) in FirebaseFirestore(FIRCollectionReference.o)
          -[FIRCollectionReference parent] in FirebaseFirestore(FIRCollectionReference.o)
          -[FIRCollectionReference path] in FirebaseFirestore(FIRCollectionReference.o)
          ...
      "___cxa_free_exception", referenced from:
          std::__1::__throw_length_error(char const*) in FirebaseFirestore(FIRDocumentSnapshot.o)
          std::__1::__throw_length_error(char const*) in FirebaseFirestore(FIRFieldPath.o)
          std::__1::__throw_length_error(char const*) in FirebaseFirestore(FIRQuery.o)
          std::__1::__throw_length_error(char const*) in FirebaseFirestore(FSTUserDataConverter.o)
          std::__1::__throw_length_error(char const*) in FirebaseFirestore(write_batch.o)
          std::__1::__throw_length_error(char const*) in FirebaseFirestore(query_snapshot.o)
          std::__1::__throw_future_error(std::__1::future_errc) in FirebaseFirestore(document_reference.o)
          ...
    ld: symbol(s) not found for architecture arm64
    clang: error: linker command failed with exit code 1 (use -v to see invocation)
    note: Using new build system
    note: Building targets in parallel
    note: Planning build
    note: Constructing build description
    Encountered error while building for device.`
    
    I use this package:   

       `firebase_messaging: ^7.0.3`
       
   This is my changes for pod file:
       
      `pod 'FirebaseFirestore', :git => 'https://github.com/invertase/firestore-ios-sdk-frameworks.git', :tag => '7.3.0'`
       
   Perhaps the reason is the version mismatch? Any advice?

**Comment by mikehardy** [comment]
If you use only firebase_messaging, you should not be interested in a firestore framework at all. You should not use this package. It would be a compile accelerant for firebase_firestore.

**Comment by FetFrumos** [comment]
What then is the reason for the very long compilation of the application? how can i speed it up?

**Comment by mikehardy** [comment]
By profiling the compilation and optimizing the longest running parts, something like this (but I want to be clear: I am not advocating any given optimization here, this is just an idea about strategy and measurement): https://medium.com/flawless-app-stories/analyzing-and-improving-build-times-in-ios-5e2b77ef408e

You have to be serious about investigating the actual build, and what takes actual time, and why, then maybe there are things you can do, just like any optimization process.

**Comment by Salakar** [comment]
Hey everyone 👋 

macOS is now supported from SDK version 7.4.0 onwards - see https://github.com/FirebaseExtended/flutterfire/pull/4698 for a full example of integrating on macOS which resulted in a roughly 8 minute reduction on the FlutterFire CI there ⚡

**Comment by muzzah** [comment]
I have to agree with some of the sentiment here, I upgraded all the version of firestore and firebase core to the latest, also specifying 7.4.0 for the tag for compiled binaries, did all the cleaning and pod updating, and still getting no where. 
Im not a paying customer yet since I am still in development mode, but as a result of wasting so much time, Im considering moving away from firebase. Id expect better support and instructions from a framework that is inherently tied to a paid service. 

Im constantly getting 

```
[!] CocoaPods could not find compatible versions for pod "cloud_firestore":
  In Podfile:
    cloud_firestore (from `.symlinks/plugins/cloud_firestore/ios`)

Specs satisfying the `cloud_firestore (from `.symlinks/plugins/cloud_firestore/ios`)` dependency were found, but they required a higher minimum deployment target.
```

and cannot fix it.

**Comment by hardikvij195** [comment]
Hi, When i add cloud_firestore dependency .......my pod install takes more than 2 hrs to finish pod install. Its get stuck on Installing BoringSSL-GRPC (0.0.7) and then again on Installing gRPC-Core (1.28.2). I am using vscode and have flutter stable channel.

Does anyone know why this happens?

**Comment by mikehardy** [comment]
Hi there! cloud_firestore is a mammoth dependency. You say "stuck", are you monitoring network traffic? It is likely not stuck, it is likely just slowly transferring data related to those dependencies.

**Comment by hardikvij195** [comment]
Hi, Thanks for replying. I am trying to connect cloud firestore with my ios app and it takes a lot of time for pod install ( around 3 hrs). What should i do to make it fast as i have a lot of projects in android and i have recently bought a macbook pro m1 and i am connecting all my projects to ios as well ? Is this issue because of m1 chip or because of latest cloud firestore dependency with flutter ? I have also attached a screenshot. 
![Screenshot 2021-02-03 at 9 19 36 PM (2)](https://user-images.githubusercontent.com/62573477/106772363-f00c4c80-6665-11eb-8f96-f284b9c049d2.png)

**Comment by suztomo** [comment]
@hardikvij195 Does the slowness happen even an empty project with just a firebase_firestore dependency?

**Comment by mikehardy** [comment]
>  are you monitoring network traffic? It is likely not stuck, it is likely just slowly transferring data related to those dependencies.

Monitor your network traffic, perhaps upgrade your network link. Once it a pod is downloaded the first time, the pod is cached so future downloads are avoided, separate from flutter, as a cocoapods thing 

You will need to experiment and learn about how cocoapods works with regards to pod fetching, pod caching, pod installation if optimizing this is important to you. It should not have anything to do with the hardware (M1 or not).

This is not something specific to cloud_firestore (other than it's reliance on firebase-ios-sdk cocoapods) so I won't be replying further sorry - it will have to do with your local network speed and disk speed.

**Comment by hardikvij195** [comment]
Hi, @mikehardy I have a 50 mbps network so i don't think its a network thing. And it happens only when i include cloud_firestore dependency.....if i only include firebase_core...pod install happens fast. Hi @suztomo Yes the slowness happens when i create a new flutter project and just include these dependency and upgrade my deployment target to ios 12.
Thanks for replying @mikehardy ...I really appreciate it.

**Comment by aradevelope** [comment]
Thank you a lot! works

**Comment by zumrywahid** [comment]
Use folliwing line for 

`  pod 'FirebaseFirestore', :git => 'https://github.com/invertase/firestore-ios-sdk-frameworks.git', :tag => '7.3.0'`

for your puspec   `cloud_firestore: "^0.16.0+1" `version .

**Comment by mikehardy** [comment]
@zumrywahid a more general recommendation (as your recommendation will go out of date as versions are released) is to consult the exact location that the firebase-ios-sdk is defined for the `firebase_core` package

That's right here: https://github.com/FirebaseExtended/flutterfire/blob/master/packages/firebase_core/firebase_core/ios/firebase_sdk_version.rb#L2

And it's possible to inspect tags / versions on that file in case you are not using the latest version

It is currently 7.3.0 as you mention though, yes :-)

**Comment by Python3-8** [comment]
I get this error:

```
[!] CocoaPods could not find compatible versions for pod "FirebaseFirestore":
  In Podfile:
    FirebaseFirestore (from `https://github.com/invertase/firestore-ios-sdk-frameworks.git`, tag `8.0.0`)

    cloud_firestore (from `.symlinks/plugins/cloud_firestore/ios`) was resolved to 2.0.0, which depends on
      Firebase/Firestore (= 7.11.0) was resolved to 7.11.0, which depends on
        FirebaseFirestore (~> 7.11.0)
```

**NOTE**: I changed the tag from `7.4.0` to `8.0.0`, because `7.4.0` didn't work either.

**Comment by bounty1342** [comment]
You should take the version matching the one  of your package. 
It screams 7.11 from the logs...

**Comment by Python3-8** [comment]
Sorry, I have absolutely no experience in iOS development. What file am I supposed to edit?

**Comment by Python3-8** [comment]
I changed this:

```ruby
pod 'FirebaseFirestore', :git => 'https://github.com/invertase/firestore-ios-sdk-frameworks.git', :tag => '8.0.0'
```

...to this:

```ruby
pod 'FirebaseFirestore', :git => 'https://github.com/invertase/firestore-ios-sdk-frameworks.git', :tag => '7.1.0'
```

...and ran `pod update`. It ran successfully and I'm gonna try building my app now.

**Comment by Python3-8** [comment]
The build still takes a long time... It took 8 minutes. Is this because it was the first build, or do I have any other dependencies that might be causing a problem?

These are my dependencies:

```yaml
dependencies:
  flutter:
    sdk: flutter
  cloud_firestore: ^2.0.0
  cupertino_icons: ^1.0.2
  email_validator: ^2.0.1
  firebase_auth: ^1.1.3
  firebase_core: ^1.1.0
  flutter_icons: ^1.1.0
  shared_preferences: ^2.0.5
  intl: ^0.17.0
  scrollable_positioned_list: ^0.1.10
  provider: ^5.0.0
  flutter_markdown: ^0.6.2
  flutter_slidable: ^0.6.0
```

**Comment by Python3-8** [comment]
I think it's actually working! I just built my app again to work on it a little bit, and the build finished in 82 seconds.

```
Launching lib/main.dart on iPhone 12 in debug mode...
lib/main.dart:1
Xcode build done.                                           81.8s
Connecting to VM Service at ws://127.0.0.1:58642/...
...
```

**Comment by Valentin-Seehausen** [comment]
> You should take the version matching the one of your package.
> It screams 7.11 from the logs...

Could this appear in the explanatory text? I stumbled upon it and I guess an explanation would be helpful to other people, too.

**Comment by mikehardy** [comment]
@Valentin-Seehausen hi there! If it is missing in some location that seems appropriate to you, I think in github you can just hit the little edit button on the README.md and suggest where it should be. We'd be happy to merge any reasonable change that would help users, most of our documentation is improved in that way since we (that is, the maintainers) are usually too closed to / used to things to have the perspective of fresh users anymore.

**Comment by kasvith** [comment]
Thank you, this worked well :)

Make sure to check the actual Firebase version in Podfile

**Comment by lemonbitsapp** [comment]
guys, im using SwiftPackage manager, - the build times is usually 40-50mins thru CI/CD. this is O M G. is there a setting to speed this up. Tks.

**Comment by mikehardy** [comment]
Depending on your CI/CD there are other tricks to speed it up since this doesn't do SPM yet
https://github.com/hendrikmuhs/ccache-action/ or
https://github.com/mikehardy/buildcache-action

Those are for github but the general solution design ("Install compiler cache tool, make symbolic links to the tool (or update your build definition to use it directly), save and restore compiler cache across builds") is generic and I have seen it work on bitrise as well so I know it's generally applicable.

**Comment by nathanaelneveux** [comment]
Trying to implement the automatic versioning based on what firebase_core uses (per [#4698](https://github.com/FirebaseExtended/flutterfire/pull/4698)) but kinda hitting a wall.

This line of course doesn't point to anything on my system - I also can't make assumptions about where it is on other members of my teams system's:
https://github.com/FirebaseExtended/flutterfire/blob/7d806d710fa6c3968373245e0b7f6bcf9dde0d96/packages/cloud_firestore/cloud_firestore/example/ios/Podfile#L41

Trying to find a replacement for this line I've come up with this:
```
def firebase_core
  generated_flutter_plugins_path = File.expand_path(File.join('..', '..', '.flutter-plugins'), __FILE__)
  unless File.exist?(generated_flutter_plugins_path)
    raise "#{generated_flutter_plugins_path} must exist. If you're running pod install manually, make sure flutter pub get is executed first"
  end

  File.foreach(generated_flutter_plugins_path) do |line|
    matches = line.match(/firebase_core\=(.*)/)
    return matches[1].strip if matches
  end
  raise "firebase_core not found in #{generated_flutter_plugins_path}. Try deleting .flutter-plugins, then run flutter pub get"
end
```
which would allow me to reference the script file like this:
```
firebase_core_script = File.expand_path(File.join('ios', 'firebase_sdk_version.rb'), firebase_core)
```

However flutter responds like this when I ty to build:
```
Exception: Warning: Podfile is out of date
  This can cause issues if your application depends on plugins that do not support iOS or macOS.
  See https://flutter.dev/docs/development/packages-and-plugins/developing-packages#plugin-platforms for details.
  If you have local Podfile edits you would like to keep, see https://github.com/flutter/flutter/issues/45197 for instructions.
```

It seems that having any mention of `.flutter-plugins` in your podfile causes this message to appear and not build. So I'm at a bit of a loss as to where to find a consistent reference to the `firebase_core` folder.

**Comment by devbrandonlong** [comment]
Went from  

<img width="597" alt="Screen Shot 2021-08-01 at 8 54 00 PM" src="https://user-images.githubusercontent.com/37224947/127791104-3c8456a7-41dd-45d4-9829-9f0d9044eb35.png">

to 

<img width="591" alt="Screen Shot 2021-08-01 at 8 54 13 PM" src="https://user-images.githubusercontent.com/37224947/127791111-8e8f3332-e430-4072-8d73-4f081351e346.png">


Nicely done 🙌🏽

**Comment by misteroak** [comment]
Yes, significant improvement. Went from 568s to 32s (MBP 3.1 GHz Quad-Core Intel Core i7). Thank you!

**Comment by dreadjr** [comment]
> Depending on your CI/CD there are other tricks to speed it up since this doesn't do SPM yet
> 
> https://github.com/hendrikmuhs/ccache-action/ or
> 
> https://github.com/mikehardy/buildcache-action
> 
> 
> 
> Those are for github but the general solution design ("Install compiler cache tool, make symbolic links to the tool (or update your build definition to use it directly), save and restore compiler cache across builds") is generic and I have seen it work on bitrise as well so I know it's generally applicable.

@mikehardy do you have a bitrise.yaml example?

**Comment by mikehardy** [comment]
No, sorry. But I've heard of a successful bitrise implementation of ccache using xcodebuild directly on command line with variables LD/LD++/CLANG/CLANG++ set on the command line to the path of ccache

**Comment by Ahmadshoh** [comment]
Thanks. It works for me. True, I had to change the version from 7.14.0 to 8.10.0, if anyone has an error like that:

[!] CocoaPods could not find compatible versions for pod "FirebaseFirestore":
  In Podfile:
    FirebaseFirestore (from `https://github.com/invertase/firestore-ios-sdk-frameworks.git`, tag `8.0.0`)

    cloud_firestore (from `.symlinks/plugins/cloud_firestore/ios`) was resolved to 2.0.0, which depends on
      Firebase/Firestore (= 7.11.0) was resolved to 7.11.0, which depends on
        FirebaseFirestore (~> 7.11.0)

**Comment by irajeshh** [comment]
It took 1hr before & 5 mins after using this.

**Comment by mikehardy** [comment]
@irajeshh if you think that's neat (and it is!) imagine integrating ccache for your entire iOS build, so that everything (not just firestore) was only built when necessary...

**Comment by gvsakhil** [comment]
Omg this is scary my build times were more than 1hr and now it went to 10-20 secs...

I followed step 4 from this url - 
https://firebase.flutter.dev/docs/firestore/overview/

**Comment by irajeshh** [comment]
> 

Yeah it's okay for me, Also I just tested the `macOs` platform, `Linux` is my main developement base.

**Comment by irajeshh** [comment]
> @irajeshh if you think that's neat (and it is!) imagine integrating ccache for your entire iOS build, so that everything (not just firestore) was only built when necessary...

Yeah it's okay for me, Also I just tested the `macOs` platform, `Linux` is my main developement base.
